File: src/unifestbestellbot/repo.py
"""Data-access functions. Each takes a Session and handles its own commit.
Audit events are written in the same transaction as the action they describe."""

import json
from dataclasses import dataclass
from datetime import date, datetime

from sqlalchemy import or_
from sqlalchemy import update as sa_update
from sqlmodel import Session, select

from .models import (
    PEER_NOTIFY_COLUMN,
    AuditEvent,
    Registration,
    Ticket,
    TicketStatus,
    now_utc,
    to_local,
)

# ---------------------------------------------------------------------------
# Registration
# ---------------------------------------------------------------------------


def upsert_registration(s: Session, reg: Registration) -> Registration:
    existing = s.get(Registration, reg.chat_id)
    if existing is not None:
        previous = existing.group_name
        existing.group_name = reg.group_name
        existing.username = reg.username
        existing.first_name = reg.first_name
        existing.last_name = reg.last_name
        s.add(existing)
        record = existing
        payload = {"group_name": reg.group_name, "previous": previous}
    else:
        s.add(reg)
        record = reg
        payload = {"group_name": reg.group_name}
    s.add(
        AuditEvent(
            kind="register",
            actor_chat_id=reg.chat_id,
            payload_json=json.dumps(payload),
        )
    )
    s.commit()
    s.refresh(record)
    return record


def unregister(s: Session, chat_id: int) -> str | None:
    """Returns the group_name the user was in, or None if nothing was deleted."""
    reg = s.get(Registration, chat_id)
    if reg is None:
        return None
    group = reg.group_name
    s.delete(reg)
    s.add(
        AuditEvent(
            kind="unregister",
            actor_chat_id=chat_id,
            payload_json=json.dumps({"group_name": group}),
        )
    )
    s.commit()
    return group


def registration_for(s: Session, chat_id: int) -> Registration | None:
    return s.get(Registration, chat_id)


def all_registrations(s: Session) -> list[Registration]:
    """Every registration, ordered by group then chat_id. For the developer
    /system snapshot; not used in any hot path."""
    stmt = select(Registration).order_by(
        Registration.group_name,
        Registration.chat_id,  # ty: ignore[invalid-argument-type]
    )
    return list(s.exec(stmt))


def group_members(
    s: Session,
    group_name: str,
    *,
    exclude_muted: bool = False,
    exclude_kind: str | None = None,
) -> list[int]:
    stmt = select(Registration.chat_id).where(Registration.group_name == group_name)
    if exclude_muted:
        now = now_utc()
        # ty/mypy can't see SQLAlchemy's column descriptor; the runtime
        # column expression supports both `.is_(None)` and `<= now`.
        mute_col = Registration.mute_peer_until
        stmt = stmt.where(
            or_(mute_col.is_(None), mute_col <= now)  # ty: ignore[unresolved-attribute, unsupported-operator]
        )
    if exclude_kind is not None:
        # Drop members who opted out of this peer-notification kind via /notify.
        kind_col = getattr(Registration, PEER_NOTIFY_COLUMN[exclude_kind])
        stmt = stmt.where(~kind_col)
    return list(s.exec(stmt))


def set_mute(s: Session, chat_id: int, *, until: datetime | None) -> bool:
    """Set or clear the peer-activity mute on a registration. Returns
    True if the row existed and was updated."""
    reg = s.get(Registration, chat_id)
    if reg is None:
        return False
    reg.mute_peer_until = until
    s.add(reg)
    s.commit()
    return True


def set_display_override(s: Session, chat_id: int, value: str | None) -> bool:
    """Set or clear a registration's self-chosen display name. Returns True
    if the row existed and was updated."""
    reg = s.get(Registration, chat_id)
    if reg is None:
        return False
    reg.display_override = value
    s.add(reg)
    s.commit()
    return True


def toggle_notify_mute(s: Session, chat_id: int, kind: str) -> bool | None:
    """Flip the per-kind peer-notification opt-out. Returns the new muted
    state (True = suppressed), or None if the registration doesn't exist."""
    reg = s.get(Registration, chat_id)
    if reg is None:
        return None
    col = PEER_NOTIFY_COLUMN[kind]
    new_state = not bool(getattr(reg, col))
    setattr(reg, col, new_state)
    s.add(reg)
    s.commit()
    return new_state


# ---------------------------------------------------------------------------
# Tickets
# ---------------------------------------------------------------------------


def create_ticket(
    s: Session,
    *,
    category: str,
    text: str,
    group_requesting: str,
    group_tasked: str,
    actor_chat_id: int,
) -> Ticket:
    t = Ticket(
        category=category,
        text=text,
        group_requesting=group_requesting,
        group_tasked=group_tasked,
    )
    s.add(t)
    s.flush()  # populate t.id
    s.add(AuditEvent(kind="open", ticket_id=t.id, actor_chat_id=actor_chat_id))
    s.commit()
    s.refresh(t)
    return t


def set_wip(s: Session, ticket_id: int, *, who: str, actor_chat_id: int) -> Ticket:
    """Atomically claim an OPEN ticket as WIP for `who`.

    The OPEN→WIP transition is a single conditional UPDATE rather than a
    read-check-write, so two near-simultaneous /wip on the same ticket can
    never both succeed: SQLite's write lock serialises them and only the
    statement that still matches `status == OPEN` flips the row. The loser
    sees rowcount 0 and gets the same ValueError as a stale claim. This is
    what makes WIP usable for task distribution (exactly one owner)."""
    t = s.get(Ticket, ticket_id)
    if t is None:
        raise LookupError(f"ticket {ticket_id} does not exist")
    stmt = (
        sa_update(Ticket)
        .where(Ticket.id == ticket_id)  # ty: ignore[invalid-argument-type]
        .where(Ticket.status == TicketStatus.OPEN)  # ty: ignore[invalid-argument-type]
        .values(status=TicketStatus.WIP, who_wip=who, who_wip_chat_id=actor_chat_id)
        .execution_options(synchronize_session=False)
    )
    # The conditional UPDATE runs on the session's connection (same
    # transaction). Going through the Connection rather than Session.execute
    # gives a CursorResult with rowcount and avoids SQLModel's select-oriented
    # exec() typing. rowcount 0 means we lost the race (someone already moved
    # it off OPEN); `t` is still OPEN here, that's fine for the message.
    if s.connection().execute(stmt).rowcount == 0:
        raise ValueError(f"ticket {ticket_id} is not open")
    s.add(
        AuditEvent(
            kind="wip",
            ticket_id=ticket_id,
            actor_chat_id=actor_chat_id,
            payload_json=json.dumps({"who": who}),
        )
    )
    s.commit()
    s.refresh(t)  # the UPDATE bypassed the ORM; reload t to reflect WIP/who
    return t


def close_ticket(s: Session, ticket_id: int, *, actor_chat_id: int) -> Ticket:
    t = s.get(Ticket, ticket_id)
    if t is None:
        raise LookupError(f"ticket {ticket_id} does not exist")
    if t.status == TicketStatus.CLOSED:
        raise ValueError(f"ticket {ticket_id} is already closed")
    t.status = TicketStatus.CLOSED
    t.closed_at = now_utc()
    s.add(t)
    s.add(AuditEvent(kind="close", ticket_id=t.id, actor_chat_id=actor_chat_id))
    s.commit()
    s.refresh(t)
    return t


def move_ticket(
    s: Session, ticket_id: int, *, new_group: str, actor_chat_id: int
) -> Ticket:
    t = s.get(Ticket, ticket_id)
    if t is None:
        raise LookupError(f"ticket {ticket_id} does not exist")
    if t.status != TicketStatus.OPEN:
        raise ValueError(f"ticket {ticket_id} is not open (status={t.status})")
    previous = t.group_tasked
    t.group_tasked = new_group
    s.add(t)
    s.add(
        AuditEvent(
            kind="move",
            ticket_id=t.id,
            actor_chat_id=actor_chat_id,
            payload_json=json.dumps({"from": previous, "to": new_group}),
        )
    )
    s.commit()
    s.refresh(t)
    return t


def get_ticket(s: Session, ticket_id: int) -> Ticket | None:
    return s.get(Ticket, ticket_id)


def ticket_count(s: Session) -> int:
    """Total tickets in the DB across all statuses. For the startup summary."""
    return len(list(s.exec(select(Ticket))))


def active_tickets(
    s: Session,
    *,
    group_tasked: str | None = None,
    status: TicketStatus | None = None,
    who_wip_chat_id: int | None = None,
) -> list[Ticket]:
    stmt = select(Ticket).where(Ticket.status != TicketStatus.CLOSED)
    if group_tasked is not None:
        stmt = stmt.where(Ticket.group_tasked == group_tasked)
    if status is not None:
        stmt = stmt.where(Ticket.status == status)
    if who_wip_chat_id is not None:
        stmt = stmt.where(Ticket.who_wip_chat_id == who_wip_chat_id)
    # Ticket.id is an InstrumentedAttribute at the class level, but its
    # declared type is `int | None`; ty/mypy can't see the descriptor magic.
    stmt = stmt.order_by(Ticket.id)  # ty: ignore[invalid-argument-type]
    return list(s.exec(stmt))


def recent_tickets(s: Session, *, limit: int = 10) -> list[Ticket]:
    """The most recently created tickets across all statuses, newest first.
    For the developer /system snapshot."""
    stmt = (
        select(Ticket)
        .order_by(Ticket.id.desc())  # ty: ignore[unresolved-attribute]
        .limit(limit)
    )
    return list(s.exec(stmt))


def closed_tickets_for_chat_id(s: Session, chat_id: int) -> list[Ticket]:
    """All CLOSED tickets the given orga worked on (claimed via /wip).
    who_wip_chat_id survives the close transition, so this is the durable
    record of a person's handled tickets — used by /self for stats."""
    stmt = (
        select(Ticket)
        .where(Ticket.status == TicketStatus.CLOSED)
        .where(Ticket.who_wip_chat_id == chat_id)
        .order_by(Ticket.id)  # ty: ignore[invalid-argument-type]
    )
    return list(s.exec(stmt))


def tickets_requested_by(s: Session, group: str) -> list[Ticket]:
    stmt = (
        select(Ticket)
        .where(Ticket.group_requesting == group)
        .where(Ticket.status != TicketStatus.CLOSED)
        .order_by(Ticket.id)  # ty: ignore[invalid-argument-type]
    )
    return list(s.exec(stmt))


def recent_tickets_for_group(s: Session, group: str, *, limit: int = 5) -> list[Ticket]:
    """The `limit` most recent tickets a group requested, newest first,
    across all statuses. Backs the /history <group> inspection view."""
    stmt = (
        select(Ticket)
        .where(Ticket.group_requesting == group)
        .order_by(Ticket.id.desc())  # ty: ignore[unresolved-attribute]
        .limit(limit)
    )
    return list(s.exec(stmt))


def recent_closed_for_group(s: Session, group: str, *, limit: int = 5) -> list[Ticket]:
    """The `limit` most recently closed tickets a group requested, newest
    first. Backs the recently-resolved section of /status."""
    stmt = (
        select(Ticket)
        .where(Ticket.group_requesting == group)
        .where(Ticket.status == TicketStatus.CLOSED)
        .order_by(Ticket.id.desc())  # ty: ignore[unresolved-attribute]
        .limit(limit)
    )
    return list(s.exec(stmt))


@dataclass
class CloseSummary:
    """One closed ticket, with metadata for display in /history."""

    ticket: Ticket
    closed_at: datetime
    closer_chat_id: int | None
    closer_display: str


def recent_closes(
    s: Session, *, group_tasked: str | None = None, limit: int = 10
) -> list[CloseSummary]:
    """Return up to `limit` most recently closed tickets, newest first.
    Each entry carries the closer's display name (from their registration
    if they still have one; otherwise their raw chat id)."""
    # SQLAlchemy comparison expressions look like `bool` to ty/mypy but
    # are ColumnElements at runtime; the join condition needs one ignore.
    stmt = (
        select(Ticket, AuditEvent)
        .join(AuditEvent, AuditEvent.ticket_id == Ticket.id)  # ty: ignore[invalid-argument-type]
        .where(Ticket.status == TicketStatus.CLOSED)
        .where(AuditEvent.kind == "close")
    )
    if group_tasked is not None:
        stmt = stmt.where(Ticket.group_tasked == group_tasked)
    stmt = stmt.order_by(AuditEvent.ts.desc()).limit(limit)  # ty: ignore[unresolved-attribute]

    summaries: list[CloseSummary] = []
    for ticket, event in s.exec(stmt):
        reg = (
            s.get(Registration, event.actor_chat_id)
            if event.actor_chat_id is not None
            else None
        )
        if reg is not None:
            display = reg.display_name()
        elif event.actor_chat_id is not None:
            display = f"chat {event.actor_chat_id}"
        else:
            display = "Unbekannt"
        summaries.append(
            CloseSummary(
                ticket=ticket,
                closed_at=event.ts,
                closer_chat_id=event.actor_chat_id,
                closer_display=display,
            )
        )
    return summaries


def record_message(
    s: Session, *, ticket_id: int, actor_chat_id: int, message: str
) -> None:
    s.add(
        AuditEvent(
            kind="message",
            ticket_id=ticket_id,
            actor_chat_id=actor_chat_id,
            payload_json=json.dumps({"message": message}),
        )
    )
    s.commit()


def record_direct_message(
    s: Session, *, chat_id: int, actor_chat_id: int, message: str
) -> None:
    """Audit a developer message sent directly to a Telegram chat."""
    s.add(
        AuditEvent(
            kind="direct_message",
            actor_chat_id=actor_chat_id,
            payload_json=json.dumps({"chat_id": chat_id, "message": message}),
        )
    )
    s.commit()


# ---------------------------------------------------------------------------
# Stats (read-only, for /stats)
# ---------------------------------------------------------------------------


@dataclass
class Percentiles:
    """Processing-time percentile summary in seconds."""

    p50_s: float
    p75_s: float
    p90_s: float
    p95_s: float
    max_s: float


@dataclass
class Stats:
    """Whole-database aggregate over every ticket seen this event. Counts are
    over all tickets regardless of status; timings are over the tickets that
    reached the relevant transition. Location is intentionally absent — it's
    config-derived, so the handler folds `by_group` into locations."""

    total: int
    open: int
    wip: int
    closed: int
    by_category: dict[str, int]
    by_group: dict[str, int]  # group_requesting -> count
    by_day: dict[date, int]  # local-time creation date -> count
    by_hour: dict[int, int]  # local-time creation hour -> count
    wait_median_s: float | None  # created -> closed, over closed tickets
    wait_max_s: float | None
    pickup_median_s: float | None  # created -> first WIP, over claimed tickets
    pickup_max_s: float | None
    wait_percentiles_by_group: dict[str, Percentiles]  # group_tasked -> created -> closed


def _median(values: list[float]) -> float | None:
    if not values:
        return None
    ordered = sorted(values)
    n = len(ordered)
    mid = n // 2
    return ordered[mid] if n % 2 else (ordered[mid - 1] + ordered[mid]) / 2


def _percentile(values: list[float], fraction: float) -> float:
    """Linearly interpolate an inclusive percentile over non-empty values."""
    ordered = sorted(values)
    index = (len(ordered) - 1) * fraction
    lower = int(index)
    upper = min(lower + 1, len(ordered) - 1)
    return ordered[lower] + (ordered[upper] - ordered[lower]) * (index - lower)


def _percentiles(values: list[float]) -> Percentiles:
    return Percentiles(
        p50_s=_percentile(values, 0.50),
        p75_s=_percentile(values, 0.75),
        p90_s=_percentile(values, 0.90),
        p95_s=_percentile(values, 0.95),
        max_s=max(values),
    )


def stats_summary(s: Session) -> Stats:
    tickets = list(s.exec(select(Ticket)))
    status = {TicketStatus.OPEN: 0, TicketStatus.WIP: 0, TicketStatus.CLOSED: 0}
    by_category: dict[str, int] = {}
    by_group: dict[str, int] = {}
    by_day: dict[date, int] = {}
    by_hour: dict[int, int] = {}
    waits: list[float] = []
    waits_by_group: dict[str, list[float]] = {}
    created_by_id: dict[int, datetime] = {}
    for t in tickets:
        status[t.status] += 1
        by_category[t.category] = by_category.get(t.category, 0) + 1
        by_group[t.group_requesting] = by_group.get(t.group_requesting, 0) + 1
        local_created = to_local(t.created_at)
        by_day[local_created.date()] = by_day.get(local_created.date(), 0) + 1
        hour = local_created.hour
        by_hour[hour] = by_hour.get(hour, 0) + 1
        if t.id is not None:
            created_by_id[t.id] = t.created_at
        if t.closed_at is not None:
            wait = (t.closed_at - t.created_at).total_seconds()
            waits.append(wait)
            waits_by_group.setdefault(t.group_tasked, []).append(wait)

    # Pickup lag: created -> first 'wip' audit event per ticket. Ordered by ts
    # ascending so the first row seen for a ticket is its earliest claim.
    first_wip: dict[int, datetime] = {}
    wip_events = s.exec(
        select(AuditEvent)
        .where(AuditEvent.kind == "wip")
        .order_by(AuditEvent.ts)  # ty: ignore[invalid-argument-type]
    )
    for e in wip_events:
        if e.ticket_id is not None and e.ticket_id not in first_wip:
            first_wip[e.ticket_id] = e.ts
    pickups = [
        (ts - created_by_id[tid]).total_seconds()
        for tid, ts in first_wip.items()
        if tid in created_by_id
    ]

    return Stats(
        total=len(tickets),
        open=status[TicketStatus.OPEN],
        wip=status[TicketStatus.WIP],
        closed=status[TicketStatus.CLOSED],
        by_category=by_category,
        by_group=by_group,
        by_day=by_day,
        by_hour=by_hour,
        wait_median_s=_median(waits),
        wait_max_s=max(waits) if waits else None,
        pickup_median_s=_median(pickups),
        pickup_max_s=max(pickups) if pickups else None,
        wait_percentiles_by_group={
            group: _percentiles(group_waits) for group, group_waits in waits_by_group.items()
        },
    )

File: rust/src/repo.rs
//! Data-access layer. Mirrors the Python `repo.py` function set; each
//! call commits its own changes and writes a matching audit-log row in
//! the same transaction.

use chrono::NaiveDateTime;
use serde_json::json;
use sqlx::{Row, SqlitePool};
use thiserror::Error;

use crate::models::{now_utc, AuditEvent, Registration, Ticket, TicketStatus};

#[derive(Debug, Error)]
pub enum RepoError {
    #[error("sqlx: {0}")]
    Sqlx(#[from] sqlx::Error),
    #[error("ticket {0} does not exist")]
    NotFound(i64),
    #[error("invalid transition: {0}")]
    InvalidTransition(String),
}

// --- Registration --------------------------------------------------------

pub async fn upsert_registration(
    pool: &SqlitePool,
    reg: &Registration,
) -> Result<Registration, RepoError> {
    let mut tx = pool.begin().await?;

    let previous: Option<String> = sqlx::query_scalar(
        "SELECT group_name FROM registration WHERE chat_id = ?",
    )
    .bind(reg.chat_id)
    .fetch_optional(&mut *tx)
    .await?;

    sqlx::query(
        r#"INSERT INTO registration
            (chat_id, group_name, username, first_name, last_name, registered_at, mute_peer_until)
           VALUES (?, ?, ?, ?, ?, ?, ?)
           ON CONFLICT(chat_id) DO UPDATE SET
             group_name = excluded.group_name,
             username   = excluded.username,
             first_name = excluded.first_name,
             last_name  = excluded.last_name"#,
    )
    .bind(reg.chat_id)
    .bind(&reg.group_name)
    .bind(&reg.username)
    .bind(&reg.first_name)
    .bind(&reg.last_name)
    .bind(reg.registered_at)
    .bind(reg.mute_peer_until)
    .execute(&mut *tx)
    .await?;

    let payload = match previous {
        Some(p) => json!({"group_name": reg.group_name, "previous": p}),
        None => json!({"group_name": reg.group_name}),
    };
    write_audit(&mut tx, "register", None, Some(reg.chat_id), Some(payload)).await?;
    tx.commit().await?;

    Ok(registration_for(pool, reg.chat_id)
        .await?
        .expect("just inserted"))
}

pub async fn unregister(pool: &SqlitePool, chat_id: i64) -> Result<Option<String>, RepoError> {
    let mut tx = pool.begin().await?;
    let group: Option<String> =
        sqlx::query_scalar("SELECT group_name FROM registration WHERE chat_id = ?")
            .bind(chat_id)
            .fetch_optional(&mut *tx)
            .await?;
    let Some(group) = group else {
        return Ok(None);
    };
    sqlx::query("DELETE FROM registration WHERE chat_id = ?")
        .bind(chat_id)
        .execute(&mut *tx)
        .await?;
    write_audit(
        &mut tx,
        "unregister",
        None,
        Some(chat_id),
        Some(json!({"group_name": group})),
    )
    .await?;
    tx.commit().await?;
    Ok(Some(group))
}

pub async fn registration_for(
    pool: &SqlitePool,
    chat_id: i64,
) -> Result<Option<Registration>, RepoError> {
    let row = sqlx::query(
        r#"SELECT chat_id, group_name, username, first_name, last_name, registered_at,
                  mute_peer_until
             FROM registration
            WHERE chat_id = ?"#,
    )
    .bind(chat_id)
    .fetch_optional(pool)
    .await?;
    Ok(row.map(row_to_registration))
}

pub async fn group_members(
    pool: &SqlitePool,
    group_name: &str,
    exclude_muted: bool,
) -> Result<Vec<i64>, RepoError> {
    let rows = if exclude_muted {
        let now = now_utc();
        sqlx::query_scalar::<_, i64>(
            r#"SELECT chat_id FROM registration
                WHERE group_name = ?
                  AND (mute_peer_until IS NULL OR mute_peer_until <= ?)"#,
        )
        .bind(group_name)
        .bind(now)
        .fetch_all(pool)
        .await?
    } else {
        sqlx::query_scalar::<_, i64>(
            "SELECT chat_id FROM registration WHERE group_name = ?",
        )
        .bind(group_name)
        .fetch_all(pool)
        .await?
    };
    Ok(rows)
}

pub async fn set_mute(
    pool: &SqlitePool,
    chat_id: i64,
    until: Option<NaiveDateTime>,
) -> Result<bool, RepoError> {
    let result = sqlx::query("UPDATE registration SET mute_peer_until = ? WHERE chat_id = ?")
        .bind(until)
        .bind(chat_id)
        .execute(pool)
        .await?;
    Ok(result.rows_affected() > 0)
}

// --- Tickets -------------------------------------------------------------

pub async fn create_ticket(
    pool: &SqlitePool,
    category: &str,
    text: &str,
    group_requesting: &str,
    group_tasked: &str,
    actor_chat_id: i64,
) -> Result<Ticket, RepoError> {
    let mut tx = pool.begin().await?;
    let created_at = now_utc();

    let result = sqlx::query(
        r#"INSERT INTO ticket
            (status, category, text, group_requesting, group_tasked, created_at)
           VALUES (?, ?, ?, ?, ?, ?)"#,
    )
    .bind(TicketStatus::Open.as_str())
    .bind(category)
    .bind(text)
    .bind(group_requesting)
    .bind(group_tasked)
    .bind(created_at)
    .execute(&mut *tx)
    .await?;

    let id = result.last_insert_rowid();
    write_audit(&mut tx, "open", Some(id), Some(actor_chat_id), None).await?;
    tx.commit().await?;

    Ok(get_ticket(pool, id).await?.expect("just inserted"))
}

pub async fn set_wip(
    pool: &SqlitePool,
    ticket_id: i64,
    who: &str,
    actor_chat_id: i64,
) -> Result<Ticket, RepoError> {
    let mut tx = pool.begin().await?;
    let ticket = fetch_ticket_in_tx(&mut tx, ticket_id).await?;
    if ticket.status != TicketStatus::Open {
        return Err(RepoError::InvalidTransition(format!(
            "ticket {} is not open (status={})",
            ticket_id, ticket.status
        )));
    }
    sqlx::query("UPDATE ticket SET status = ?, who_wip = ? WHERE id = ?")
        .bind(TicketStatus::Wip.as_str())
        .bind(who)
        .bind(ticket_id)
        .execute(&mut *tx)
        .await?;
    write_audit(
        &mut tx,
        "wip",
        Some(ticket_id),
        Some(actor_chat_id),
        Some(json!({"who": who})),
    )
    .await?;
    tx.commit().await?;
    Ok(get_ticket(pool, ticket_id).await?.expect("just updated"))
}

pub async fn close_ticket(
    pool: &SqlitePool,
    ticket_id: i64,
    actor_chat_id: i64,
) -> Result<Ticket, RepoError> {
    let mut tx = pool.begin().await?;
    let ticket = fetch_ticket_in_tx(&mut tx, ticket_id).await?;
    if ticket.status == TicketStatus::Closed {
        return Err(RepoError::InvalidTransition(format!(
            "ticket {} is already closed",
            ticket_id
        )));
    }
    let closed_at = now_utc();
    sqlx::query("UPDATE ticket SET status = ?, closed_at = ? WHERE id = ?")
        .bind(TicketStatus::Closed.as_str())
        .bind(closed_at)
        .bind(ticket_id)
        .execute(&mut *tx)
        .await?;
    write_audit(&mut tx, "close", Some(ticket_id), Some(actor_chat_id), None).await?;
    tx.commit().await?;
    Ok(get_ticket(pool, ticket_id).await?.expect("just updated"))
}

pub async fn move_ticket(
    pool: &SqlitePool,
    ticket_id: i64,
    new_group: &str,
    actor_chat_id: i64,
) -> Result<Ticket, RepoError> {
    let mut tx = pool.begin().await?;
    let ticket = fetch_ticket_in_tx(&mut tx, ticket_id).await?;
    if ticket.status != TicketStatus::Open {
        return Err(RepoError::InvalidTransition(format!(
            "ticket {} is not open (status={})",
            ticket_id, ticket.status
        )));
    }
    let previous = ticket.group_tasked.clone();
    sqlx::query("UPDATE ticket SET group_tasked = ? WHERE id = ?")
        .bind(new_group)
        .bind(ticket_id)
        .execute(&mut *tx)
        .await?;
    write_audit(
        &mut tx,
        "move",
        Some(ticket_id),
        Some(actor_chat_id),
        Some(json!({"from": previous, "to": new_group})),
    )
    .await?;
    tx.commit().await?;
    Ok(get_ticket(pool, ticket_id).await?.expect("just updated"))
}

pub async fn get_ticket(pool: &SqlitePool, ticket_id: i64) -> Result<Option<Ticket>, RepoError> {
    let row = sqlx::query(
        r#"SELECT id, status, category, text, group_requesting, group_tasked,
                  who_wip, created_at, closed_at
             FROM ticket
            WHERE id = ?"#,
    )
    .bind(ticket_id)
    .fetch_optional(pool)
    .await?;
    Ok(row.map(row_to_ticket))
}

pub async fn active_tickets(
    pool: &SqlitePool,
    group_tasked: Option<&str>,
    status: Option<TicketStatus>,
) -> Result<Vec<Ticket>, RepoError> {
    let base = r#"SELECT id, status, category, text, group_requesting, group_tasked,
                         who_wip, created_at, closed_at
                    FROM ticket
                   WHERE status != ?"#;
    let mut q = String::from(base);
    if group_tasked.is_some() {
        q.push_str(" AND group_tasked = ?");
    }
    if status.is_some() {
        q.push_str(" AND status = ?");
    }
    q.push_str(" ORDER BY id");

    let mut query = sqlx::query(&q).bind(TicketStatus::Closed.as_str());
    if let Some(g) = group_tasked {
        query = query.bind(g);
    }
    if let Some(s) = status {
        query = query.bind(s.as_str());
    }
    let rows = query.fetch_all(pool).await?;
    Ok(rows.into_iter().map(row_to_ticket).collect())
}

pub async fn tickets_requested_by(
    pool: &SqlitePool,
    group: &str,
) -> Result<Vec<Ticket>, RepoError> {
    let rows = sqlx::query(
        r#"SELECT id, status, category, text, group_requesting, group_tasked,
                  who_wip, created_at, closed_at
             FROM ticket
            WHERE group_requesting = ? AND status != ?
            ORDER BY id"#,
    )
    .bind(group)
    .bind(TicketStatus::Closed.as_str())
    .fetch_all(pool)
    .await?;
    Ok(rows.into_iter().map(row_to_ticket).collect())
}

pub async fn record_message(
    pool: &SqlitePool,
    ticket_id: i64,
    actor_chat_id: i64,
    message: &str,
) -> Result<(), RepoError> {
    let mut tx = pool.begin().await?;
    write_audit(
        &mut tx,
        "message",
        Some(ticket_id),
        Some(actor_chat_id),
        Some(json!({"message": message})),
    )
    .await?;
    tx.commit().await?;
    Ok(())
}

// --- Recent closes (for /history) ----------------------------------------

#[derive(Debug, Clone)]
pub struct CloseSummary {
    pub ticket: Ticket,
    pub closed_at: NaiveDateTime,
    pub closer_chat_id: Option<i64>,
    pub closer_display: String,
}

pub async fn recent_closes(
    pool: &SqlitePool,
    group_tasked: Option<&str>,
    limit: i64,
) -> Result<Vec<CloseSummary>, RepoError> {
    let mut q = String::from(
        r#"SELECT t.id, t.status, t.category, t.text, t.group_requesting, t.group_tasked,
                  t.who_wip, t.created_at, t.closed_at,
                  e.ts AS event_ts, e.actor_chat_id AS event_actor
             FROM ticket t
             JOIN auditevent e ON e.ticket_id = t.id
            WHERE t.status = ? AND e.kind = ?"#,
    );
    if group_tasked.is_some() {
        q.push_str(" AND t.group_tasked = ?");
    }
    q.push_str(" ORDER BY e.ts DESC LIMIT ?");

    let mut query = sqlx::query(&q)
        .bind(TicketStatus::Closed.as_str())
        .bind("close");
    if let Some(g) = group_tasked {
        query = query.bind(g);
    }
    query = query.bind(limit);

    let rows = query.fetch_all(pool).await?;

    let mut out = Vec::with_capacity(rows.len());
    for row in rows {
        let event_ts: NaiveDateTime = row.try_get("event_ts")?;
        let event_actor: Option<i64> = row.try_get("event_actor")?;
        let ticket = row_to_ticket(row);
        let closer_display = match event_actor {
            Some(cid) => match registration_for(pool, cid).await? {
                Some(reg) => reg.display_name(),
                None => format!("chat {}", cid),
            },
            None => "Unbekannt".to_string(),
        };
        out.push(CloseSummary {
            ticket,
            closed_at: event_ts,
            closer_chat_id: event_actor,
            closer_display,
        });
    }
    Ok(out)
}

// --- Audit listing (for tests) -------------------------------------------

pub async fn all_audit(pool: &SqlitePool) -> Result<Vec<AuditEvent>, RepoError> {
    let rows = sqlx::query(
        "SELECT id, ts, kind, ticket_id, actor_chat_id, payload_json FROM auditevent ORDER BY id",
    )
    .fetch_all(pool)
    .await?;
    Ok(rows
        .into_iter()
        .map(|r| AuditEvent {
            id: r.get("id"),
            ts: r.get("ts"),
            kind: r.get("kind"),
            ticket_id: r.get("ticket_id"),
            actor_chat_id: r.get("actor_chat_id"),
            payload_json: r.get("payload_json"),
        })
        .collect())
}

// --- Internals -----------------------------------------------------------

async fn fetch_ticket_in_tx(
    tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>,
    ticket_id: i64,
) -> Result<Ticket, RepoError> {
    let row = sqlx::query(
        r#"SELECT id, status, category, text, group_requesting, group_tasked,
                  who_wip, created_at, closed_at
             FROM ticket
            WHERE id = ?"#,
    )
    .bind(ticket_id)
    .fetch_optional(&mut **tx)
    .await?;
    row.map(row_to_ticket).ok_or(RepoError::NotFound(ticket_id))
}

async fn write_audit(
    tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>,
    kind: &str,
    ticket_id: Option<i64>,
    actor_chat_id: Option<i64>,
    payload: Option<serde_json::Value>,
) -> Result<(), RepoError> {
    let payload_json = payload.map(|v| v.to_string());
    sqlx::query(
        "INSERT INTO auditevent (ts, kind, ticket_id, actor_chat_id, payload_json)
         VALUES (?, ?, ?, ?, ?)",
    )
    .bind(now_utc())
    .bind(kind)
    .bind(ticket_id)
    .bind(actor_chat_id)
    .bind(payload_json)
    .execute(&mut **tx)
    .await?;
    Ok(())
}

fn row_to_registration(row: sqlx::sqlite::SqliteRow) -> Registration {
    Registration {
        chat_id: row.get("chat_id"),
        group_name: row.get("group_name"),
        username: row.get("username"),
        first_name: row.get("first_name"),
        last_name: row.get("last_name"),
        registered_at: row.get("registered_at"),
        mute_peer_until: row.get("mute_peer_until"),
    }
}

fn row_to_ticket(row: sqlx::sqlite::SqliteRow) -> Ticket {
    let status_str: String = row.get("status");
    let status = TicketStatus::parse(&status_str).expect("known status");
    Ticket {
        id: row.get("id"),
        status,
        category: row.get("category"),
        text: row.get("text"),
        group_requesting: row.get("group_requesting"),
        group_tasked: row.get("group_tasked"),
        who_wip: row.get("who_wip"),
        created_at: row.get("created_at"),
        closed_at: row.get("closed_at"),
    }
}
